### PR TITLE
Ipv6 Support - IPv6 compatibility refactor

### PR DESCRIPTION
**Pull request by nlgwcy**
**What type of PR is this?**

Add one of the following kinds:
/kind feature

**What this PR does / why we need it**:
As [issues/291 comment](https://github.com/kmesh-net/kmesh/issues/291#issuecomment-2169836525), This PR covers only IPv6 compatibility reconstruction.
1. workload mode supports IPv6
   - [x] IPv6-compatible reconstruction on the control plane
   - [x] IPv6-compatible reconstruction on the data plane

**Which issue(s) this PR fixes**:
Fixes https://github.com/kmesh-net/kmesh/issues/291

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
```

self test:
1. fortio test

   ```sh
   [root@localhost images]# kubectl get svc
   NAME                     TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)               AGE
   fortio-server            ClusterIP   10.96.250.180   <none>        80/TCP                2d12h
   
   [root@localhost images]# kubectl get endpoints
   NAME                     ENDPOINTS         AGE
   default-istio-waypoint                     102s
   fortio-server            10.244.1.4:9090   2d12h
   
   # fortio test
   [root@localhost kmesh]# kubectl exec -ti fortio-client-deployment-6c86f967bf-rxwrb -- fortio load -c 1 -t 20s http://10.96.250.180:80
   
   # ss -antp
   [root@localhost images]# nsenter -t 617877 -n ss -antp
   State                  Recv-Q              Send-Q                           Local Address:Port                            Peer Address:Port              Process
   TIME-WAIT              0                   0                                  10.244.1.12:46496                             10.244.1.4:9090
   ESTAB                  0                   0                                  10.244.1.12:45562                             10.244.1.4:9090               users:(("fortio",pid=617877,fd=3))
   ```

2. waypoint

   ```sh
   # fortio test
   [root@localhost images]# kubectl exec -ti fortio-client-deployment-6c86f967bf-xtf2b -- fortio load -c 1 -t 30s http://10.96.153.141:80
   [root@localhost httpbin]# nsenter -t 943363 -n ss -antp
   State               Recv-Q              Send-Q                           Local Address:Port                              Peer Address:Port               Process
   ESTAB               0                   0                                  10.244.1.14:54676                               10.244.1.9:15019               users:(("fortio",pid=943363,fd=3))
   ```

3. auth

   - pods info

     ```sh
     # pods
     [root@localhost httpbin]# kubectl get pods -owide
     NAME                                        READY   STATUS    RESTARTS   AGE     IP            NODE             NOMINATED NODE   READINESS GATES
     fortio-client-deployment-6c86f967bf-xtf2b   1/1     Running   0          7m13s   10.244.1.14   ambient-worker   <none>           <none>
     fortio-server-deployment-7fcf96565c-s22p2   1/1     Running   0          73m     10.244.1.4    ambient-worker   <none>           <none>
     httpbin-f85dc8498-wp226                     1/1     Running   0          4m4s    10.244.1.17   ambient-worker   <none>           <none>
     sleep-78ff5975c6-w46dv                      1/1     Running   0          4m5s    10.244.1.16   ambient-worker   <none>           <none>
     sleep-78ff5975c6-xrhn7                      1/1     Running   0          4m5s    10.244.1.15   ambient-worker   <none>           <none>
     # svc
     [root@localhost httpbin]# kubectl get svc
     NAME            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
     fortio-server   ClusterIP   10.96.153.141   <none>        80/TCP     73m
     httpbin         ClusterIP   10.96.149.231   <none>        8000/TCP   4m7s
     kubernetes      ClusterIP   10.96.0.1       <none>        443/TCP    75m
     sleep           ClusterIP   10.96.80.90     <none>        80/TCP     4m9s
     [root@localhost httpbin]
     ```

   - config accept auth rule

     ```yaml
     kubectl apply -f - <<EOF
     apiVersion: security.istio.io/v1beta1
     kind: AuthorizationPolicy
     metadata:
      name: allow-by-srcip
      namespace: default
     spec:
      selector:
        matchLabels:
          app: httpbin
      action: ALLOW
      rules:
      - from:
        - source:
            ipBlocks:
            - 10.244.1.15/32
     EOF
     ```

   - test

     ```sh
     # curl 
     [root@localhost httpbin]# kubectl exec -ti sleep-78ff5975c6-w46dv -- curl http://10.96.149.231:8000/headers
     curl: (56) Recv failure: Connection reset by peer
     command terminated with exit code 56
     [root@localhost httpbin]# kubectl exec -ti sleep-78ff5975c6-xrhn7 -- curl http://10.96.149.231:8000/headers
     {
       "headers": {
         "Accept": "*/*",
         "Host": "10.96.149.231:8000",
         "User-Agent": "curl/8.8.0"
       }
     }
     ```

   - config deny auth rule

     ```yaml
     kubectl apply -f - <<EOF
     apiVersion: security.istio.io/v1beta1
     kind: AuthorizationPolicy
     metadata:
      name: allow-by-srcip
      namespace: default
     spec:
      selector:
        matchLabels:
          app: httpbin
      action: ALLOW
      rules:
      - from:
        - source:
            ipBlocks:
            - 10.244.1.15/32
     EOF
     ```

   - test

     ```sh
     [root@localhost httpbin]# kubectl exec -ti sleep-78ff5975c6-xrhn7 -- curl http://10.96.149.231:8000/headers
     curl: (56) Recv failure: Connection reset by peer
     command terminated with exit code 56
     [root@localhost httpbin]# kubectl exec -ti sleep-78ff5975c6-w46dv -- curl http://10.96.149.231:8000/headers
     {
       "headers": {
         "Accept": "*/*",
         "Host": "10.96.149.231:8000",
         "User-Agent": "curl/8.8.0"
       }
     }
     [root@localhost httpbin]#
     ```

### CODE REVIEW

File: bpf/include/bpf_common.h [hunk 1]
```diff
@@ -1,5 +1,9 @@
 /* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
 /* Copyright Authors of Kmesh */
+#ifndef __KMESH_BPF_COMMON_H__
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_common.h:3]: nit: add a blank line

File: bpf/include/bpf_common.h [hunk 2]
```diff
@@ -12,6 +16,13 @@
 /*0x3a4(932) is the specific port handled by the daemon to enable bypass*/
 #define DISABLE_BYPASS_PORT 0x3a4
 
+struct manager_key {
+    union {
+        __u64 netns_cookie;
+        struct ip_addr addr;
```
Comment (hzxuzhonghu) [inline bpf/include/bpf_common.h:22]: why add a address type? when is it used

File: bpf/include/common.h [hunk 3]
```diff
@@ -36,6 +36,24 @@
...
 
+struct ip_addr {
+    union {
+        __u32 ip4;
+        __u32 ip6[4];
+    };
+};
+#define IPV6_ADDR_LEN 16
+
+struct ctx_info {
```
Comment (hzxuzhonghu) [inline bpf/include/common.h:47]: please add a comment about this struct

Comment (hzxuzhonghu) [reply bpf/include/common.h:47]: What does it mean, and when is it used?

File: bpf/kmesh/workload/cgroup_sock.c [hunk 4]
```diff
@@ -4,20 +4,21 @@
...
+static inline int sock4_traffic_control(struct bpf_sock_addr *ctx, struct ctx_info *info)
 {
     int ret;
     frontend_value *frontend_v = NULL;
 
     if (ctx->protocol != IPPROTO_TCP)
         return 0;
 
-    DECLARE_FRONTEND_KEY(ctx, frontend_k);
+    DECLARE_FRONTEND_KEY(ctx, &info->vip, frontend_k);
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/cgroup_sock.c:21]: Cannot we get the v6 addr from ctx?

File: bpf/kmesh/workload/cgroup_sock.c [hunk 5]
```diff
@@ -48,8 +54,19 @@ int cgroup_connect4_prog(struct bpf_sock_addr *ctx)
         return CGROUP_SOCK_OK;
     }
 
-    int ret = sock4_traffic_control(ctx);
+    int ret = sock4_traffic_control(ctx, &info);
+    if (ret) {
+        BPF_LOG(ERR, KMESH, "sock_traffic_control failed:%d\n", ret);
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/cgroup_sock.c:59]:
```suggestion
        BPF_LOG(ERR, KMESH, "sock_traffic_control failed: %d", ret);
```

File: bpf/kmesh/workload/include/backend.h [hunk 6]
```diff
@@ -15,45 +15,49 @@ static inline backend_value *map_lookup_backend(const backend_key *key)
...
     struct bpf_sock_tuple value_tuple = {0};
 
-    value_tuple.ipv4.daddr = ctx->user_ip4;
-    value_tuple.ipv4.dport = ctx->user_port;
-
+    if (ctx->family == AF_INET) {
+        value_tuple.ipv4.daddr = info->vip.ip4;
+        value_tuple.ipv4.dport = ctx->user_port;
+    } else {
+        bpf_memcpy(value_tuple.ipv6.daddr, info->vip.ip6, IPV6_ADDR_LEN);
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/include/backend.h:29]: should check INET6

File: bpf/include/bpf_common.h (hunk 1 above)
Comment (nlgwcy) [reply bpf/include/bpf_common.h:4]: done

File: bpf/kmesh/workload/include/backend.h [hunk 7]
```diff
@@ -15,49 +15,58 @@ static inline backend_value *map_lookup_backend(const backend_key *key)
...
 
-    // if tail call failed will run this code
-    BPF_LOG(ERR, BACKEND, "workload tail call failed, err is %d\n", ret);
-    return -ENOEXEC;
+    if (ctx->user_family == AF_INET)
+        kmesh_ctx->dnat_ip.ip4 = wp_addr->ip4;
+    else
+        bpf_memcpy(kmesh_ctx->dnat_ip.ip6, wp_addr->ip6, IPV6_ADDR_LEN);
+    kmesh_ctx->dnat_port = port;
+    kmesh_ctx->via_waypoint = true;
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/include/backend.h:47]: if we move this tail call into this function, it seems we donot need `via_waypoint`

File: bpf/kmesh/workload/include/backend.h [hunk 8]
```diff
@@ -15,49 +15,58 @@ static inline backend_value *map_lookup_backend(const backend_key *key)
...
+static inline int
+backend_manager(struct kmesh_context *kmesh_ctx, backend_value *backend_v, __u32 service_id, service_value *service_v)
 {
     int ret;
-    address_t target_addr;
+    ctx_buff_t *ctx = (ctx_buff_t *)kmesh_ctx->ctx;
     __u32 user_port = ctx->user_port;
 
-    if (backend_v->waypoint_addr != 0 && backend_v->waypoint_port != 0) {
+    if (backend_v->wp_addr.ip4 != 0 && backend_v->waypoint_port != 0) {
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/include/backend.h:58]: where is the code to process ipv6

File: bpf/kmesh/workload/include/backend.h [hunk 9]
```diff
@@ -71,15 +80,18 @@ static inline int backend_manager(ctx_buff_t *ctx, backend_value *backend_v, __u
...
+                        bpf_memcpy(kmesh_ctx->dnat_ip.ip6, backend_v->addr.ip6, IPV6_ADDR_LEN);
+                    kmesh_ctx->dnat_port = service_v->target_port[j];
+                    kmesh_ctx->via_waypoint = false;
                     BPF_LOG(
                         DEBUG,
                         BACKEND,
                         "get the backend addr=[%pI4h:%u]",
-                        &target_addr.ipv4,
-                        bpf_ntohs(target_addr.port));
+                        &kmesh_ctx->dnat_ip.ip4,
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/include/backend.h:93]: this log can only fit for ipv4 too

File: bpf/kmesh/workload/sendmsg.c [hunk 10]
```diff
@@ -30,14 +31,43 @@
...
+[dst_ip4]   - 4 bytes
+[dst_port]  - 2 bytes
+*/
+#define TLV_ORG_DST_ADDR4_LENGTH (TLV_IP4_LENGTH + TLV_PORT_LENGTH)
+#define TLV_ORG_DST_ADDR6_LENGTH (TLV_IP6_LENGTH + TLV_PORT_LENGTH)
+
+/*
+[TYPE]      - 1 byte
+[length]    - 4 bytes
+[dst_ip4]   - TLV_ORG_DST_ADDR4_LENGTH bytes
```
Comment (hzxuzhonghu) [inline bpf/kmesh/workload/sendmsg.c:48]: TLV_IP4_LENGTH

File: bpf/kmesh/workload/include/backend.h (hunk 7 above)
Comment (nlgwcy) [reply bpf/kmesh/workload/include/backend.h:47]: Before tail call, need dnat dstIp(v4/v6), bpf verifier will fail.

File: bpf/kmesh/workload/include/backend.h (hunk 8 above)
Comment (nlgwcy) [reply bpf/kmesh/workload/include/backend.h:58]: I missed it, It looks like we only need to check if waypoint_port is zero.

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

File: .clang-format [hunk 11]
```diff
@@ -82,6 +82,8 @@ BraceWrapping:
   BeforeCatch: false
   # else之前
   BeforeElse: false
+  # elseif之前
+  BeforeElseif: false
```
Comment (hzxuzhonghu) [inline .clang-format:86]: This is breaking

**Pull request merged**